Sources/NIOExtras/LengthFieldBasedFrameDecoder.swift

```diff
@@ -0,0 +1,170 @@
+///
+/// Given that the specified header length is 1 byte,
+/// where the first header specifies 3 bytes while the second header specifies 4 bytes,
+/// a `LengthFieldBasedFrameDecoder` will decode them into the following packets:
+///
+///     +-----+------+
+///     | BCD | FGHI |
+///     +-----+------+
+///
+/// 'A' and 'E' will be the headers and will not be passed forward.
+///
+
+public final class LengthFieldBasedFrameDecoder: ByteToMessageDecoder {
+    
+    ///
+    /// The decoder has two distinct sections of data to read.
+    /// Each must be fully present before it is considered as read.
+    /// During the time when it is not present the decoder must wait. `DecoderReadState` details that waiting state.
+    ///
+    private enum DecoderReadState {
+        case waitingForHeader
+        case waitingForFrame(length: Int)
+    }
+
+    public typealias InboundIn = ByteBuffer
+    public typealias InboundOut = ByteBuffer
+    
+    public var cumulationBuffer: ByteBuffer?
+    private var readState: DecoderReadState = .waitingForHeader
+    
+    private let lengthFieldLength: ByteLength
+    private let lengthFieldEndianness: Endianness
+    
+    /// Create `LengthFieldBasedFrameDecoder` with a given frame length.
+    ///
+    /// - parameters:
+    ///    - lengthFieldLength: The length of the field specifying the remaining length of the frame.
+    ///    - lengthFieldEndianness: The endianness of the field specifying the remaining length of the frame.
+    ///
+    public init(lengthFieldLength: ByteLength, lengthFieldEndianness: Endianness = .big) {
+
+        // The value contained in the length field must be able to be represented by an integer type on the platform.
+        // ie. .eight == 64bit which would not fit into the Int type on a 32bit platform.
+        precondition(lengthFieldLength.length <= Int.bitWidth/8)
+            
+        self.lengthFieldLength = lengthFieldLength
+        self.lengthFieldEndianness = lengthFieldEndianness
+    }
+    
+    public func decode(ctx: ChannelHandlerContext, buffer: inout ByteBuffer) throws -> DecodingState {
+        
+        if case .waitingForHeader = self.readState {
+            try self.readNextLengthFieldToState(buffer: &buffer)
+        }
+        
+        guard case .waitingForFrame(let frameLength) = self.readState else {
+            return .needMoreData
+        }
+        
+        guard let frameBuffer = try self.readNextFrame(buffer: &buffer, frameLength: frameLength) else {
+            return .needMoreData
+        }
+        
+        ctx.fireChannelRead(self.wrapInboundOut(frameBuffer))
+
+        return .continue
+    }
+    
+    private func readNextLengthFieldToState(buffer: inout ByteBuffer) throws {
+        
+        guard let lengthFieldSlice = buffer.readSlice(length: self.lengthFieldLength.length) else {
+            return
+        }
+        
+        // Convert the length field to an integer specifying the length
```

Comment:
this still can't be right. You're mutating `buffer` (by reading something off) but you're not mutating `self.readState`.

Generally, `self.readState` will need to remember all changes you have already done where there's a way to return in the mean time. So now, you'll have an issue (and this needs a test case) if you receive the bytes in this order:

- `channelRead`: just a `5`
- now we read off a slice `lengthFieldSlice` (correct)
- then we try to do `self.frameLength` but if that fails, we'll return having read `lengthFieldSlice` off already but we'll still be in the `.waitingForHeader` state (incorrect)

So if we get bytes again, say `1, 2, 3, 4, 5` (which should be the data), we'll interpret the `1` as a new header...

Reply:
Okay, so I am going to write a test where the data come in and is read with single bytes, so that if there is an error by either splitting the header or the data it will catch it.

Reply:
So, I split the test into single bytes. It didn't catch this error. I think because it is actually very difficult to get the first call to succeed and the second to fail. As mentioned I neatened it out anyway as it is better not to have the call twice.